src/sentry/similarity.py

```diff
@@ -0,0 +1,245 @@
+      data can be used to identify what buckets a key is a member of, and also
+      used to identify the degree of bucket similarity when comparing with data
+      associated with another key.
+    - A bucket membership set, which maintains a record of what keys have been
+      record in a ``(band, bucket)`` pair. This data can be used to identify
+      what other keys may be similar to the lookup key (but not the degree of
+      similarity.)
+    """
+    BUCKET_MEMBERSHIP = '0'
+    BUCKET_FREQUENCY = '1'
+
+    def __init__(self, cluster, rows, bands, buckets):
+        self.namespace = b'sim'
+
+        self.cluster = cluster
+        self.rows = rows
+
+        generator = random.Random(0)
+
+        def shuffle(value):
+            generator.shuffle(value)
+            return value
+
+        self.bands = [
+            [shuffle(range(rows)) for _ in xrange(buckets)]
+            for _ in xrange(bands)
+        ]
+
+        self.__bucket_formatter = get_number_formatter(rows)
+
+    def __format_buckets(self, bucket):
+        return b''.join(
+            map(self.__bucket_formatter, bucket)
+        )
+
+    def get_signature(self, value):
+        """Generate a minhash signature for a value."""
+        columns = set(hash(token) % self.rows for token in value)
+        return map(
+            lambda band: map(
+                lambda permutation: next(i for i, a in enumerate(permutation) if a in columns),
+                band,
+            ),
+            self.bands,
+        )
+
+    def get_similarity(self, target, other):
+        """\
+        Calculate the degree of similarity between two bucket frequency
+        sequences which represent two different keys.
+
+        This is mainly an implementation detail for sorting query results, but
+        is exposed publically for testing. This method assumes all input
+        values have already been normalized using ``scale_to_total``.
+        """
+        assert len(target) == len(other)
+        assert len(target) == len(self.bands)
+        return 1 - sum(
+            map(
+                lambda (left, right): get_distance(
+                    left,
+                    right,
+                ) / math.sqrt(2),
+                zip(target, other)
+            )
+        ) / len(target)
+
+    def query(self, scope, key):
+        """\
+        Find other entries that are similar to the one repesented by ``key``.
+
+        This returns an sequence of ``(key, estimated similarity)`` pairs,
+        where a similarity score of 1 is completely similar, and a similarity
+        score of 0 is completely dissimilar. The result sequence is ordered
+        from most similar to least similar. (For example, the search key itself
+        isn't filtered from the result and will always have a similarity of 1,
+        typically making it the first result.)
+        """
+        def fetch_bucket_frequencies(keys):
```

Comment:
Long-winded comment incoming!

It's not *necessary* per se (though `fetch_bucket_frequencies` is called twice, so that one felt like natural move.) You could absolutely get the same behavior without these two inner functions, but I often do this in long or complex functions for a few reasons:

- It makes it a lot easier to understand the overall control flow [from lines 207-230](https://github.com/getsentry/sentry/blob/5b044a8924b9056a58868c6072e2c7fa18de3a5d/src/sentry/similarity.py#L207-L230) by being able to refer to these "subroutines" that are defined up front by name. My hope is that when you're first reading the code, you skim over these functions to get an idea of what they're taking as inputs and providing as outputs. Then, when you're later reading where they're actually used, you can be more concerned with *what* the program is doing (fetching bucket frequencies for keys) rather than *how* it's doing it (querying Redis, resolving promises, normalizing bucket values, etc.) If you then have questions about the specific implementation while you're reading the code where they're used, you can then double back and brush up on the implementation details.
-  It decreases the likelihood of unintentional name shadowing in what would otherwise be a reasonably long function, especially for names like `responses`, `key`, or `band` here. Since there are fewer local variables in play in each independent function, this also makes it easier to understand. It also decreases the number of local variables each frame in case this ends up in a Sentry stacktrace, which can help make debugging.
- By *not* making these methods on the class, it effectively makes them part of a private interface, and also signals that they're only intended to be used for this specific purpose in this method. (I have a personal vendetta against public methods that are only intended to be called from different methods in within the same class definition, and have no practical reason to be overridden.)